### PR TITLE
drivers: sensor: adxl372: Updated driver with RTIO stream functionality

### PR DESCRIPTION
**Pull request by vladislav-pejic**
Updated ADXL372 driver with RTIO stream functionality. 
RTIO stream is using both FIFO threshold and FIFO full triggers. 
Together with RTIO stream, RTIO async read is also implemented.

Signed-off-by: Vladislav Pejic vladislav.pejic@orioninc.com

### CODE REVIEW

File: drivers/sensor/adi/adxl372/adxl372_rtio.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	}
+}
+
+void adxl372_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	struct rtio_work_req *req = rtio_work_req_alloc();
+
+	__ASSERT_NO_MSG(req);
+
+	rtio_work_req_submit(req, iodev_sqe, adxl372_submit_helper);
```
Comment (teburd) [inline drivers/sensor/adi/adxl372/adxl372_rtio.c:64]: Is the work queue usage really needed here? Ideally you'd use RTIO with the spi bus as well to allow for the gpio ISR to directly start a spi transfer and avoid the context swaps.

**Review by teburd — COMMENTED** [review]
Looks like a really nice driver, very happy to see another device taking advantage of this functionality. Now I need to get an analog adxl372 on my bench it seems 💯

File: drivers/sensor/adi/adxl372/adxl372.h [hunk 2]
```diff
@@ -358,9 +376,23 @@ struct adxl372_dev_config {
...
+	bool is_fifo;
+	uint64_t timestamp;
+	uint8_t int_status;
+	uint8_t sample_set_size: 5;
+	uint8_t has_x: 1;
+	uint8_t has_y: 1;
+	uint8_t has_z: 1;
+	uint16_t accel_odr: 4;
+	uint16_t fifo_byte_count: 12;
+} __attribute__((__packed__));
```
[the remark this answers is not in the document]
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372.h:390]: I agree and changes are in the latest code. I needed to do some rearranging to align it and should be ok now.

File: drivers/sensor/adi/adxl372/adxl372_decoder.c [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+ * Copyright (c) 2024 Analog Devices Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "adxl372.h"
+
+#ifdef CONFIG_ADXL372_STREAM
+
+static uint32_t accel_period_ns[] = {
```
[the remark this answers is not in the document]
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372_decoder.c]: Yes. Changes are in the latest code.

File: drivers/sensor/adi/adxl372/adxl372_decoder.c [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+	[ADXL372_ODR_400HZ] = UINT32_C(1000000000) / 400,
+	[ADXL372_ODR_800HZ] = UINT32_C(1000000000) / 800,
+	[ADXL372_ODR_1600HZ] = UINT32_C(100000000) / 1600,
+	[ADXL372_ODR_3200HZ] = UINT32_C(100000000) / 3200,
+	[ADXL372_ODR_6400HZ] = UINT32_C(100000000) / 6400,
+};
+
+static void adxl372_accel_convert_q31(q31_t *out, uint16_t data_in)
```
[the remark this answers is not in the document]
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372_decoder.c]: Can you check the latest code to see if this is what you had in mind?

File: drivers/sensor/adi/adxl372/adxl372_decoder.c [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+	return ret;
+}
+
+static int adxl372_decode_sample(const struct adxl372_xyz_accel_data *data,
+		struct sensor_chan_spec chan_spec, uint32_t *fit, uint16_t max_count, void *data_out)
+{
+	struct sensor_value *out = (struct sensor_value *) data_out;
+
+    if (*fit > 0) {
```
[the remark this answers is not in the document]
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372_decoder.c]: Fixed

File: drivers/sensor/adi/adxl372/adxl372_stream.c [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+	uint16_t fifo_samples = (((data->fifo_ent[0] & 0x3) << 8) | data->fifo_ent[1]);
+	size_t sample_set_size = adxl372_get_packet_size(cfg);
+
+	/* At least one sample set must remain in FIFO to encure that data
+	 * is not overwritten and stored out of order.
+	 */
+	if (fifo_samples > sample_set_size/2) {
+		fifo_samples -= sample_set_size/2;
+	} else {
+		LOG_ERR("fifo sample caunt error %d\n", fifo_samples);
```
Comment (teburd) [inline drivers/sensor/adi/adxl372/adxl372_stream.c:118]: s/caunt/count I think? or caught?

File: drivers/sensor/adi/adxl372/adxl372_stream.c [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+		packet_size = 2;
+		break;
+
+	case ADXL372_XY_FIFO:
+	case ADXL372_XZ_FIFO:
+	case ADXL372_YZ_FIFO:
+		packet_size = 4;
+		break;
+
+	default:
```
Comment (teburd) [inline drivers/sensor/adi/adxl372/adxl372_stream.c:96]: packet_size = 6 (in the default case) here would be clearer than initializing it above imho

File: drivers/sensor/adi/adxl372/adxl372_stream.c [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+	if ((cfg->fifo_config.fifo_format == ADXL372_Z_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_YZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XYZ_FIFO))	{
+		hdr.has_z = 1;
+	}
+
+	uint32_t buf_avail = buf_len;
+
+	memcpy(buf, &hdr, sizeof(hdr));
```
Comment (teburd) [inline drivers/sensor/adi/adxl372/adxl372_stream.c:181]: Could avoid the memcpy here perhaps by treating the buffer as a *hdr and initializing from there, though that requires that everything is word aligned.

File: drivers/sensor/adi/adxl372/adxl372_stream.c [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+	((struct adxl372_fifo_data *)buf)->fifo_byte_count = read_len;
+
+	__ASSERT_NO_MSG(read_len % pkt_size == 0);
+
+	uint8_t *read_buf = buf + sizeof(hdr);
+
+	/* Flush completions */
+	struct rtio_cqe *cqe;
+
+	do {
```
Comment (teburd) [inline drivers/sensor/adi/adxl372/adxl372_stream.c:198]: It's likely worth checking error codes here and bailing with rtio_iodev_sqe_err() on a bus error if one occurs.

**Review by teburd — COMMENTED** [review]
Looks pretty good, some minor things I noticed and commented on

File: drivers/sensor/adi/adxl372/adxl372_stream.c (hunk 6 above)
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372_stream.c]: Fixed

File: drivers/sensor/adi/adxl372/adxl372_stream.c (hunk 8 above)
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372_stream.c]: Done

File: drivers/sensor/adi/adxl372/adxl372_stream.c (hunk 9 above)
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372_stream.c:197]: Can you give me more information? rtio_cqe_consume returns NULL when there is no CQE and that is cowered, and rtio_cqe_release is void func. 
Did you mean for this part: 
struct rtio_sqe *write_fifo_addr = rtio_sqe_acquire(data->rtio_ctx); ?

Comment (teburd) [reply drivers/sensor/adi/adxl372/adxl372_stream.c:197]: When there is a completion queue event from the bus request, its result may be -errno, e.g.

```c
int res = 0;

do {
    cqe = rtio_cqe_consume(...);
    if (cqe != NULL) {
        if(cqe->result < 0 && res == 0) {
            /* handle spi/i2c bus error here */
            LOG_ERR("first bus error: %d", cqe->result);
            res = cqe->result;
        }
        rtio_cqe_release(cqe);
    }
} while (cqe != NULL);

/* Bail/cancel attempt to read sensor on any error */
if (res != 0) {
   rtio_iodev_sqe_err(...);
   return;
}
```

Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372_stream.c:202]: That is a good idea. It is added in the latest code.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by MaureenHelm** [comment]
Added `DNM` label at @vladislav-pejic 's request. He found an issue when `spi_transceive_dt` and `rtio_submit` are used at the same time, and will push an update on Monday.

**Comment by teburd** [comment]
> Added `DNM` label at @vladislav-pejic 's request. He found an issue when `spi_transceive_dt` and `rtio_submit` are used at the same time, and will push an update on Monday.

If one path is still using a semaphore to serialize access to the spi controller, while the other path is using the operation queue that would indeed cause issues! I didn't look too closely at this in the other PR but would be my first suspicion.

**Review by teburd — COMMENTED** [review]
Not blocking on my part, but some general commentary.

Would love to see the q conversions distilled to the single multiple they should be rather than the modulus and division currently being done.

**Comment by vladislav-pejic** [comment]
> Added `DNM` label at @vladislav-pejic 's request. He found an issue when `spi_transceive_dt` and `rtio_submit` are used at the same time, and will push an update on Monday.

Fix pushed in the latest commit.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by vladislav-pejic** [comment]
> > > Added `DNM` label at @vladislav-pejic 's request. He found an issue when `spi_transceive_dt` and `rtio_submit` are used at the same time, and will push an update on Monday.
> > 
> > 
> > If one path is still using a semaphore to serialize access to the spi controller, while the other path is using the operation queue that would indeed cause issues! I didn't look too closely at this in the other PR but would be my first suspicion.
> 
> Can we get more details on this? I'm interested to know if it's present in #78346 or in main (SPI RTIO shim?)

I am using driver from that PR for my testing. I have experienced unstable behavior when calling spi_transceive_dt from RTIO callbacks. Some spi_transceive_dt are done correctly, but some are not. The ones that are not executed correctly, system will go to idle state, like is waiting for something and nothing will happen, no IRQs, nothing. It is really hard to debug because each breakpoint or debug message will change timings. This is only present when SPI_IRQ is used and was not present when old SPI driver had blocking RTIO.

File: drivers/sensor/adi/adxl372/Kconfig [hunk 10]
```diff
@@ -85,6 +86,7 @@ config ADXL372_REFERENCED_ACTIVITY_DETECTION_MODE
 
 choice ADXL372_TRIGGER_MODE
 	prompt "Trigger mode"
+	default ADXL372_TRIGGER_GLOBAL_THREAD if ADXL372_STREAM
```
[the remark this answers is not in the document]
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/Kconfig:89]: It is not. It was left from some older version. Will be fixed in the latest push.

File: drivers/sensor/adi/adxl372/Kconfig [hunk 11]
```diff
@@ -104,6 +106,15 @@ config ADXL372_TRIGGER_OWN_THREAD
 
 endchoice
 
+config ADXL372_STREAM
+	bool "Use FIFO to stream data"
+	select ADXL372_TRIGGER
```
[the remark this answers is not in the document]
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/Kconfig:111]: ADXL372_TRIGGER define is used in the driver to initialize IRQ. If it is not present, no IRQ initialization will be done. I didn't want to change original driver. In the latest push, no thread is used when RTIO is enabled, just IRQ enabling.

File: drivers/sensor/adi/adxl372/adxl372_decoder.c [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+	[ADXL372_ODR_3200HZ] = UINT32_C(100000000) / 3200,
+	[ADXL372_ODR_6400HZ] = UINT32_C(100000000) / 6400,
+};
+
+static inline void adxl372_accel_convert_q31(q31_t *out, const uint8_t *buff)
+{
+	uint16_t data_in = (*buff << 4) | ((*(buff + 1) & 0xF0) >> 4);
+	int32_t micro_ms2 = data_in * SENSOR_SCALING_FACTOR;
+
+	*out = CLAMP((((int64_t)micro_ms2) + (micro_ms2 % 1000000)), INT32_MIN, INT32_MAX);
```
[the remark this answers is not in the document]
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372_decoder.c]: Line 27 is just for rearranging the bytes due to the way how is read from the FIFO. SENSOR_SCALING_FACTOR is made from the conversion done in the original driver
`/*
	 * Sensor resolution is 100mg/LSB, 12-bit value needs to be right
	 * shifted by 4 or divided by 16. Overall this results in a scale of 160
	 */
	int32_t micro_ms2 = value * (SENSOR_G / (16 * 1000 / 100));`

I agree that this is not the most optimized way, but to be honest I still do not fully understand the conversion that is done in the original driver (adxl372_accel_convert). I would need more time to understand it. Is it ok to go like this for this PR and maybe in the future new PR can be done where both adxl372_accel_convert and q31 format?

File: drivers/sensor/adi/adxl372/adxl372.c [hunk 13]
```diff
@@ -642,12 +647,16 @@ static int adxl372_channel_get(const struct device *dev,
 }
 
 static const struct sensor_driver_api adxl372_api_funcs = {
-	.attr_set     = adxl372_attr_set,
+	.attr_set = adxl372_attr_set,
 	.sample_fetch = adxl372_sample_fetch,
-	.channel_get  = adxl372_channel_get,
+	.channel_get = adxl372_channel_get,
 #ifdef CONFIG_ADXL372_TRIGGER
 	.trigger_set = adxl372_trigger_set,
```
[the remark this answers is not in the document]
Comment (vladislav-pejic) [reply drivers/sensor/adi/adxl372/adxl372.c:691]: I thought about this at the beginning, but I went with approach were both can be used at the same time. This is how driver is working at the moment. You can choose which one do you want to use.

**Review by MaureenHelm — APPROVED** [review]

File: drivers/sensor/adi/adxl372/adxl372_decoder.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (teburd) [reply drivers/sensor/adi/adxl372/adxl372_decoder.c]: Closer...

So as you noted this device has a 100mg/lsb, meaning the scale is actually

-204.8 G to 204.7G which allows for the -200 to 200G reading range.

The sensitivity value in the linked issue math here is 10, a every count of 10 is 1G

To convert this to q31 we need a scaling multiplier that also converts to m/s^2 (200*9.8 = 1960) with a scale of -1960 to 1960, so -2048 to 2048 is the next power of 2 up. This is a shift of 11 (pow(2,11 = 2048)

So lets do our math in the issue to get a scaling factor and see if it works out.

(1.0/10.0)*(pow(2,31)/pow(2,11))*9.8 = 1027604.4800000001, which sadly isn't an exact integer, but our closest rounding is 1027604

Lets try it out... at 200G we'd have 

200G = 2000 reading value

2000*1027604 = 2055208000

To get back to m/s^2 we need divide by what amounts to 1 m/s^2 which is pow(2, 31 - shift) or pow (2,20)

2055208000/pow(2,20) = 1959 m/s^2

converting back to G

1959/9.8 = 199.89 G

So your scaling value is 1027604 works with pretty minimal error, though not a perfect scaling.

File: drivers/sensor/adi/adxl372/adxl372_decoder.c [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+{
+	int16_t data_in = ((int16_t)*buff << 4) | (((int16_t)*(buff + 1) & 0xF0) >> 4);
+
+	if (data_in & BIT(11)) {
+		data_in |= ADXL372_COMPLEMENT;
+	}
+
+	int64_t micro_ms2 = data_in * SENSOR_SCALING_FACTOR;
+
+	*out = CLAMP((micro_ms2 + (micro_ms2 % 1000000)), INT32_MIN, INT32_MAX);
```
Comment (ubieda) [inline drivers/sensor/adi/adxl372/adxl372_decoder.c:36]: I'm struggling to understand this conversion: I'd expect the output to be q32_t (with its value and shift) and be in m/s^2.

However, I don't see where:
- The shift value in `struct sensor_three_axis_data` is set (zero by default?).
- Where the value is taken from its current unit (micro-m/s^2?) to m/s^2.

Reference:
https://github.com/zephyrproject-rtos/zephyr/blob/69b73ebb0ab9b85fd2f219fa6ca1e135ba58ac88/drivers/sensor/tdk/icm42688/icm42688_decoder.c#L80-L124

May I get some clarification?

File: drivers/sensor/adi/adxl372/adxl372_decoder.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (ubieda) [reply drivers/sensor/adi/adxl372/adxl372_decoder.c]: I missed this when Ieft [the other comment](https://github.com/zephyrproject-rtos/zephyr/pull/78941#discussion_r1811265047), it seems to match with my confusion.

@teburd @MaureenHelm would you be ok with getting this PR in and then iterate on the decoder in a separate PR?

[the remark this answers is not in the document]
Comment (MaureenHelm) [reply drivers/sensor/adi/adxl372/adxl372_decoder.c]: Yes, I think that's the best way to go.

**Review by ubieda — APPROVED** [review]
+1 to get this in. We can iterate on decoding separatedly (https://github.com/zephyrproject-rtos/zephyr/pull/78941#discussion_r1811277602)

**Review by teburd — APPROVED** [review]
I'm ok with this going in, but I'd like to see some bugs filed for the decoding of adxl drivers to correctly do q31 conversions or some quick follow ups.

File: drivers/sensor/adi/adxl372/adxl372_decoder.c (hunk 14 above)
Comment (teburd) [reply drivers/sensor/adi/adxl372/adxl372_decoder.c:36]: icm42688 actually also has bugs in its conversion, which is why I created #76968 and @bperseghetti has open #67207 #76861

**Pull request merged**